### PR TITLE
gh-101046: Fix a potential memory leak in the parser when raising MemoryError

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-101046

### CODE REVIEW

File: Tools/peg_generator/pegen/c_generator.py [hunk 1]
```diff
@@ -801,7 +801,7 @@ def handle_alt_loop(self, node: Alt, is_gather: bool, rulename: Optional[str]) -
                 self.print(
                     "void **_new_children = PyMem_Realloc(_children, _children_capacity*sizeof(void *));"
                 )
-                self.out_of_memory_return(f"!_new_children")
+                self.out_of_memory_return(f"!new_children", cleanup_code="PyMem_Free(_children);")
```
Comment (kumaraditya303) [inline Tools/peg_generator/pegen/c_generator.py]:
```suggestion
                self.out_of_memory_return(f"!_new_children", cleanup_code="PyMem_Free(_children);")
```

**Review by lysnikolaou — APPROVED** [review]
LGTM!

**Comment by hugovk** [comment]
> Sorry, @pablogsal, I could not cleanly backport this to `3.9` due to a conflict. Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line. `cherry_picker 1de4395f62bb140563761ef5cbdf46accef3c550 3.9`

@pablogsal Does this still need backporting to 3.9? Otherwise let's remove the backport label.

**Comment by pablogsal** [comment]
Removed

**Pull request merged**